server/upload.go

```diff
@@ -170,3 +138,23 @@ func uploadBlobChunked(ctx context.Context, requestURL *url.URL, layer *Layer, r
 	}
 	return nil
 }
+
+type ProgressWriter struct {
+	status    string
+	digest    string
+	completed int
+	total     int
+	fn        func(api.ProgressResponse)
+}
+
+func (pw *ProgressWriter) Write(p []byte) (n int, err error) {
+	pw.completed += len(p)
+	pw.fn(api.ProgressResponse{
+		Status:    pw.status,
+		Digest:    pw.digest,
+		Total:     pw.total,
+		Completed: pw.completed,
+	})
+
+	return 0, nil
```

Comment:
does returning 0 here indicate nothing was written?

Reply:
Good catch. It should be len(p) though it doesn't seem to make a difference.